internal/controllers/core/togglebutton/reconciler.go

```diff
@@ -0,0 +1,291 @@
+		lastClickProcessTimes: make(map[string]time.Time),
+	}
+}
+
+func indexToggleButton(obj client.Object) []indexer.Key {
+	var result []indexer.Key
+	toggleButton := obj.(*v1alpha1.ToggleButton)
+	bGVK := v1alpha1.SchemeGroupVersion.WithKind("ConfigMap")
+
+	if toggleButton != nil {
+		if toggleButton.Spec.StateSource.ConfigMap != nil {
+			result = append(result, indexer.Key{
+				Name: types.NamespacedName{Name: toggleButton.Spec.StateSource.ConfigMap.Name},
+				GVK:  bGVK,
+			})
+		}
+	}
+	return result
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	nn := request.NamespacedName
+
+	var tb v1alpha1.ToggleButton
+	err := r.ctrlClient.Get(ctx, nn, &tb)
+	r.indexer.OnReconcile(nn, &tb)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+
+	if apierrors.IsNotFound(err) || !tb.ObjectMeta.DeletionTimestamp.IsZero() {
+		err := r.managedOwnedUIButton(ctx, nn, nil)
+		return ctrl.Result{}, err
+	}
+
+	err = r.processClick(ctx, tb)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = r.manageToggleButton(ctx, &tb)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = r.managedOwnedUIButton(ctx, nn, &tb)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, err
+}
+
+func (r *Reconciler) processClick(ctx context.Context, tb v1alpha1.ToggleButton) error {
+	uiButton := v1alpha1.UIButton{}
+	err := r.ctrlClient.Get(ctx, types.NamespacedName{Name: uibuttonName(tb.Name)}, &uiButton)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		} else {
+			return err
+		}
+	}
+
+	// if there's a new click, pass the new value through to the ConfigMap
+	if uiButton.Status.LastClickedAt.After(r.lastClickProcessTimes[tb.Name]) {
+		foundInput := false
+		var newOnValue bool
+		for _, input := range uiButton.Status.Inputs {
+			if input.Name == actionUIInputName {
+				if input.Hidden == nil {
+					return fmt.Errorf("button %q input %q was not of type 'Hidden'", uiButton.Name, input.Name)
+				}
+				switch input.Hidden.Value {
+				case turnOnInputValue:
+					newOnValue = true
+				case turnOffInputValue:
+					newOnValue = false
+				default:
+					return fmt.Errorf("button %q input %q had unexpected value %q", uiButton.Name, input.Name, input.Hidden.Value)
+				}
+				foundInput = true
+				break
+			}
+		}
+
+		if !foundInput {
+			return fmt.Errorf("UIButton %q does not have an input named %q", uiButton.Name, actionUIInputName)
+		}
+
+		ss := tb.Spec.StateSource.ConfigMap
+		if ss == nil {
+			return fmt.Errorf("ToggleButton %q has nil Spec.StateSource.ConfigMap", tb.Name)
+		}
+		var cm v1alpha1.ConfigMap
+		err := r.ctrlClient.Get(ctx, types.NamespacedName{Name: ss.Name}, &cm)
+		if err != nil {
+			return errors.Wrap(err, "fetching ToggleButton ConfigMap")
```

Comment:
who's responsible for creating the ConfigMap? if there's no ConfigMap, should ToggleButton create one?

(probably want to handle NotFound separately from other errors)